// vanillajs-datepicker/js/lib/dom.js
const i = document.createRange();
function s(t) {
  return i.createContextualFragment(t);
}
function o(t) {
  return t.parentElement || (t.parentNode instanceof ShadowRoot ? t.parentNode.host : void 0);
}
function r(t) {
  return t.getRootNode().activeElement === t;
}
function p(t) {
  t.style.display !== "none" && (t.style.display && (t.dataset.styleDisplay = t.style.display), t.style.display = "none");
}
function f(t) {
  t.style.display === "none" && (t.dataset.styleDisplay ? (t.style.display = t.dataset.styleDisplay, delete t.dataset.styleDisplay) : t.style.display = "");
}
function n(t) {
  t.firstChild && (t.removeChild(t.firstChild), n(t));
}
function y(t, e) {
  n(t), e instanceof DocumentFragment ? t.appendChild(e) : typeof e == "string" ? t.appendChild(s(e)) : typeof e.forEach == "function" && e.forEach((a) => {
    t.appendChild(a);
  });
}
export {
  n as emptyChildNodes,
  o as getParent,
  p as hideElement,
  r as isActiveElement,
  s as parseHTML,
  y as replaceChildNodes,
  f as showElement
};
