// vanillajs-datepicker/js/lib/event.js
const s = /* @__PURE__ */ new WeakMap(), { addEventListener: a, removeEventListener: f } = EventTarget.prototype;
function c(r, t) {
  let n = s.get(r);
  n || (n = [], s.set(r, n)), t.forEach((e) => {
    a.call(...e), n.push(e);
  });
}
function u(r) {
  let t = s.get(r);
  !t || (t.forEach((n) => {
    f.call(...n);
  }), s.delete(r));
}
if (!Event.prototype.composedPath) {
  const r = (t, n = []) => {
    n.push(t);
    let e;
    return t.parentNode ? e = t.parentNode : t.host ? e = t.host : t.defaultView && (e = t.defaultView), e ? r(e, n) : n;
  };
  Event.prototype.composedPath = function() {
    return r(this.target);
  };
}
function o(r, t, n) {
  const [e, ...i] = r;
  if (t(e))
    return e;
  if (!(e === n || e.tagName === "HTML" || i.length === 0))
    return o(i, t, n);
}
function p(r, t) {
  const n = typeof t == "function" ? t : (e) => e instanceof Element && e.matches(t);
  return o(r.composedPath(), n, r.currentTarget);
}
export {
  p as findElementInEventPath,
  c as registerListeners,
  u as unregisterListeners
};
