// lit-element/lit-element.js
import { ReactiveElement as s } from "../@lit/reactive-element/reactive-element.js";
import { ReactiveElement as v, defaultConverter as C, notEqual as E } from "../@lit/reactive-element/reactive-element.js";
import { render as d, noChange as a } from "../lit-html/lit-html.js";
import { _$LH as b, html as g, noChange as R, nothing as $, render as _ } from "../lit-html/lit-html.js";
/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
var o, r;
class n extends s {
  constructor() {
    super(...arguments), this.renderOptions = { host: this }, this._$Do = void 0;
  }
  createRenderRoot() {
    var e, t;
    const l = super.createRenderRoot();
    return (e = (t = this.renderOptions).renderBefore) !== null && e !== void 0 || (t.renderBefore = l.firstChild), l;
  }
  update(e) {
    const t = this.render();
    this.hasUpdated || (this.renderOptions.isConnected = this.isConnected), super.update(e), this._$Do = d(t, this.renderRoot, this.renderOptions);
  }
  connectedCallback() {
    var e;
    super.connectedCallback(), (e = this._$Do) === null || e === void 0 || e.setConnected(!0);
  }
  disconnectedCallback() {
    var e;
    super.disconnectedCallback(), (e = this._$Do) === null || e === void 0 || e.setConnected(!1);
  }
  render() {
    return a;
  }
}
n.finalized = !0, n._$litElement$ = !0, (o = globalThis.litElementHydrateSupport) === null || o === void 0 || o.call(globalThis, { LitElement: n });
const i = globalThis.litElementPolyfillSupport;
i == null || i({ LitElement: n });
((r = globalThis.litElementVersions) !== null && r !== void 0 ? r : globalThis.litElementVersions = []).push("3.2.2");
export {
  n as LitElement,
  v as ReactiveElement,
  b as _$LH,
  C as defaultConverter,
  g as html,
  R as noChange,
  E as notEqual,
  $ as nothing,
  _ as render
};
